### PR TITLE
Derive slug field for demos in `metadata.json`

### PR DESCRIPTION
**Pull request by Mandrenkov**
**Summary:**
This PR adds a `slug` field to each item in `metadata.json`. The slug is derived from the filename of the demo and will eventually replace the `canonicalURL` field. In a subsequent PR, the canonical URL field will be removed.

**Relevant references:**
None.

**Possible Drawbacks:**
None.

**Related GitHub Issues:**
None.

**Other Notes:**
* [x] This PR includes a new GHA workflow "Build Website - Text (Test)" for testing purposes. The workflow should be deleted prior to merging this PR.

**Verification:**
The generated `metadata.json` file has the correct `slug` field for each demo in the Glass Onion bucket.

<details><summary><tt>metadata.json</tt> Snippet</summary>
<p>

```json
[
  {
    "authors": [
      {
        "username": "whatsis"
      }
    ],
    "basedOnPapers": [],
    "canonicalURL": "/qml/demos/tutorial_how_to_use_noise_models",
    "categories": [
      "Quantum Computing",
      "How-to"
    ],
    "dateOfLastModification": "2024-10-01T00:00:00+00:00",
    "dateOfPublication": "2024-10-01T00:00:00+00:00",
    "doi": "",
    "previewImages": [
      {
        "type": "thumbnail",
        "uri": "/_static/demo_thumbnails/regular_demo_thumbnails/thumbnail_how_to_use_noise_models.png"
      },
      {
        "type": "large_thumbnail",
        "uri": "/_static/demo_thumbnails/large_demo_thumbnails/thumbnail_large_how_to_use_noise_models.png"
      }
    ],
    "referencedByPapers": [],
    "references": [],
    "relatedContent": [
      {
        "id": "tutorial_noisy_circuits",
        "type": "demonstration",
        "weight": 1
      }
    ],
    "seoDescription": "Learn how noise models can be built and inserted into a quantum circuit in PennyLane.",
    "slug": "tutorial_how_to_use_noise_models",
    "tags": [],
    "title": "How to use noise models in PennyLane"
  },
  {
    "authors": [
      {
        "username": "alvaro"
      }
    ],
    "basedOnPapers": [],
    "canonicalURL": "/qml/demos/tutorial_photonics",
    "categories": [
      "Quantum Hardware",
      "Quantum Computing"
    ],
    "dateOfLastModification": "2024-08-06T00:00:00+00:00",
    "dateOfPublication": "2022-05-31T00:00:00+00:00",
    "doi": "",
    "previewImages": [
      {
        "type": "thumbnail",
        "uri": "/_static/demo_thumbnails/regular_demo_thumbnails/thumbnail_photonic_QC.png"
      }
    ],
    "referencedByPapers": [],
    "references": [
      {
        "authors": "D. DiVincenzo",
        "id": "DiVincenzo2000",
        "journal": "Fortschritte der Physik",
        "title": "The Physical Implementation of Quantum Computation",
        "type": "article",
        "url": "https://onlinelibrary.wiley.com/doi/10.1002/1521-3978(200009)48:9/11%3C771::AID-PROP771%3E3.0.CO;2-E",
        "year": "2000"
      },
      {
        "authors": "C. Weedbrook, et al.",
        "id": "Weedbrook2012",
        "journal": "Rev. Mod. Phys.",
        "title": "Gaussian Quantum Information",
        "type": "article",
        "url": "https://journals.aps.org/rmp/abstract/10.1103/RevModPhys.84.621",
        "year": "2012"
      },
      {
        "authors": "S. Sabouri, et al.",
        "id": "Sabouri2021",
        "journal": "IEEE Photonics Journal",
        "number": "2",
        "title": "Thermo Optical Phase Shifter With Low Thermal Crosstalk for SOI Strip Waveguide",
        "type": "article",
        "url": "https://ieeexplore.ieee.org/document/9345963",
        "volume": "13",
        "year": "2021"
      },
      {
        "authors": "M. Paris",
        "id": "Paris1996",
        "journal": "Physics Letters A",
        "title": "Displacement operator by beam splitter",
        "type": "article",
        "url": "https://www.sciencedirect.com/science/article/abs/pii/0375960196003398?via%3Dihub",
        "year": "1996"
      },
      {
        "authors": "S. Braunstein, P. van Loock",
        "id": "Braunstein2005",
        "journal": "Rev. Mod. Phys.",
        "title": "Quantum information with continuous variables",
        "type": "article",
        "url": "https://journals.aps.org/rmp/abstract/10.1103/RevModPhys.77.513",
        "year": "2005"
      },
      {
        "authors": "C. Hamilton, et al.",
        "id": "Hamilton2017",
        "journal": "Phys. Rev. Lett.",
        "title": "Gaussian Boson Sampling",
        "type": "article",
        "url": "https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.119.170501",
        "year": "2017"
      },
      {
        "authors": "H.S. Zhong, et al.",
        "id": "Zhong2020",
        "journal": "Science",
        "title": "Quantum computational advantage using photons",
        "type": "article",
        "url": "https://www.science.org/doi/10.1126/science.abe8770",
        "year": "2020"
      },
      {
        "authors": "L. Madsen, et al.",
        "id": "Madsen2020",
        "journal": "Nature",
        "title": "Quantum computational advantage with a programmable photonic processor",
        "type": "article",
        "url": "https://www.nature.com/articles/s41586-022-04725-x",
        "year": "2022"
      },
      {
        "authors": "I. Tzitrin, et al.",
        "id": "Tzitrin2020",
        "journal": "Phys. Rev. A",
        "title": "Progress towards practical qubit computation using approximate Gottesman-Kitaev-Preskill codes",
        "type": "article",
        "url": "https://journals.aps.org/pra/abstract/10.1103/PhysRevA.64.012310",
        "year": "2020"
      },
      {
        "authors": "E. Bourassa, et al.",
        "id": "Bourassa2021",
        "journal": "Quantum",
        "title": "Blueprint for a Scalable Photonic Fault-Tolerant Quantum Computer",
        "type": "article",
        "url": "https://quantum-journal.org/papers/q-2021-02-04-392/",
        "year": "2021"
      }
    ],
    "relatedContent": [
      {
        "id": "tutorial_pasqal",
        "type": "demonstration",
        "weight": 1
      },
      {
        "id": "tutorial_trapped_ions",
        "type": "demonstration",
        "weight": 1
      },
      {
        "id": "tutorial_sc_qubits",
        "type": "demonstration",
        "weight": 1
      },
      {
        "id": "gbs",
        "type": "demonstration",
        "weight": 1
      }
    ],
    "seoDescription": "Learn how photonic quantum computers work through code",
    "slug": "tutorial_photonics",
    "tags": [],
    "title": "Photonic quantum computers"
  },
  {
    "authors": [
      {
        "id": "diksha_dhawan"
      }
    ],
    "basedOnPapers": [],
    "canonicalURL": "/qml/demos/tutorial_mapping",
    "categories": [
      "Algorithms",
      "Quantum Computing",
      "Quantum Chemistry"
    ],
    "dateOfLastModification": "2024-08-05T00:00:00+00:00",
    "dateOfPublication": "2024-05-06T00:00:00+00:00",
    "doi": "",
    "previewImages": [
      {
        "type": "thumbnail",
        "uri": "/_static/demonstration_assets/mapping/thumbnail_mapping_2024-06-20.png"
      },
      {
        "type": "large_thumbnail",
        "uri": "/_static/demo_thumbnails/large_demo_thumbnails/thumbnail_large_mapping_2024-06-20.png"
      }
    ],
    "referencedByPapers": [],
    "references": [
      {
        "authors": "A. Tranter, S. Sofia et al.",
        "id": "Tranter",
        "publisher": "International Journal of Quantum Chemistry",
        "title": "The Bravyi–Kitaev Transformation: Properties and Applications",
        "type": "article",
        "url": "https://onlinelibrary.wiley.com/doi/10.1002/qua.24969",
        "year": "2015"
      },
      {
        "authors": "Y. S. Yordanov et al.",
        "id": "Yordanov",
        "publisher": "Physical Review A",
        "title": "Efficient quantum circuits for quantum computational chemistry",
        "type": "article",
        "url": "https://journals.aps.org/pra/abstract/10.1103/PhysRevA.102.062612",
        "year": "2020"
      }
    ],
    "relatedContent": [],
    "seoDescription": "Learn how to map fermionic operators to qubit operators",
    "slug": "tutorial_mapping",
    "tags": [],
    "title": "Mapping fermionic Hamiltonians to qubit Hamiltonians"
  },
  # ...
]
```

</p>
</details>

### CODE REVIEW

File: metadata_schemas/demo.metadata.schema.0.1.3.json [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        "type": "string",
+        "pattern": "^$|^10[.]"
+      },
+      "canonicalURL": {
+        "description": "The canonical URL for the demo. Sometimes there might be more than one URL that points to a given page on a website. The canonical URL defines which of these should be thought of as the primary or main one.",
+        "type": "string",
+        "pattern": "(^https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6})?\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)"
+      },
+      "references": {
+        "description": "An array of the references used for the demo.",
```
Comment (anthayes92) [inline metadata_schemas/demo.metadata.schema.0.1.3.json:96]: ### Question:
Is the assumption that arrays cannot be empty unless specified?

**Review by anthayes92 — APPROVED** [review]
Looks good, appreciate the test! :test_tube:

File: metadata_schemas/demo.metadata.schema.0.1.3.json (hunk 1 above)
Comment (Mandrenkov) [reply metadata_schemas/demo.metadata.schema.0.1.3.json:96]: I'm not sure. With the exception of the `slug` field, everything else in this file was copied directly from `demo.metadata.schema.0.1.2.json`. But whatever the semantics are, they work for all existing demos.

File: .github/workflows/upload-text.yml [hunk 2]
```diff
@@ -51,11 +51,21 @@ jobs:
...
+            find demonstrations -name "*.metadata.json" -type f \
+            ! -name "demonstrations_categories.metadata.json" \
+          )
+
+          for filepath in $filepaths; do
+              slug=$(basename $filepath .metadata.json)
+              jq ". += {\"slug\": \"$slug\"}" < $filepath >> $tmp_metadata_json
+          done
+
+          jq --slurp --sort-keys . < $tmp_metadata_json > $out_metadata_json
```
Comment (doctorperceptron) [inline .github/workflows/upload-text.yml:68]: Nice!

**Review by doctorperceptron — APPROVED** [review]

**Pull request merged**